### DISCUSSION TITLE
Saving the compiled FSM by serializing it?

### DISCUSSION THREAD

**Discussion by markandpathak** [timbray/quamina#297] [Q&A] [answered]
Thank you for this fantastic library. I have a question about storing the compiled FSM and reusing it between invocation. Suppose I have few hundred rules/patterns which doesn't change much over time. But my service is ephemeral because it is hosted in lambda. 

I want to serialize the compiled instance of FSM and store it in some DB so that on each invocation of the service, it doesn't need to recompile the rules into FSM and it can get same instance by de-serializing it from stored data. Can you provide any direction how can I achieve this?

**Comment by timbray** [comment]
Good question.  First thing, have you measured the latency/compute required to build the FSM?  It's possible that it's not even significant. (I genuinely don't know, haven't really done any serious analysis of FSM-building.). If it really is a noticeable problem do please file an issue.

Let’s assume that it is significant. Then yes, it would obviously be a good idea to serialize (oops, we say “marshal” in Go) a constructed Quamina instance in a way that allowed fast restore. Which doesn't seem like rocket science.  Go has existing serializers for data structures into JSON and CBOR and various other things; once again, I don't know what the efficiency story is. But it does seem like a thing that could be done without too much effort.

**Comment by markandpathak** [comment]
I haven't measured the FSM compilation performance, but the logic is: it'll be always faster to fetch a single binary blob from a very fast key-value store like Dynamo or ScyllaDB or Redis than fetching a hundred rules in raw format from the database. So even if compilation doesn't make much difference but retrieving rules, in most of the cases will be slower than the compiled FSM binary. Anyways, I will try to marshal 😀 it with `encoding/gob` package and let you know.

**Comment by timbray** [comment, marked as answer]
I can't argue with your logic, and my experience running the analogous event-ruler package at AWS agrees with it.  

I had never previously heard of encoding/gob, I'll be curious to hear about your experience with it. If the thing works well, please consider doing a PR to build this into Quamina.

**Reply by markandpathak** [reply]
Sure, I'll let you know how it goes.

**Comment by markandpathak** [comment]
@timbray I am trying to serialize the struct, but the biggest issue I face is that all of the fields of structures are unexported. This is the case for majority of structs. This makes all of serialization very difficult. Either the struct fields needs to be exported or a serialize method needs to be implemented in all struct

**Reply by timbray** [reply]
I don't have any problem in principle with exporting the fields you need to serialize.  Would you like to do a PR that would export the ones you need?  Then we'll get input from all the contributors. 

Question: Would this involve bringing in a dependency? Quamina being dependency-free is a big advantage.

**Reply by markandpathak** [reply]
That'll be very helpful if the community can help export the fields. I tried a little, but I need to understand the code structure first to assess its effect. 
Coming to your question:
I don't think that it will involve any external dependencies. Mostly standard libraries provided in Go are sufficient. Still, I will keep that in mind to avoid external dependencies.

There can be some possibilities:
1. Even after exporting the fields, serialization may fail, which can be mainly due to data referenced through pointers. Although `encoding/gob` package handles pointers, still there can be some instances. 
2. In that case we will need to write custom serialization functions, wherever serialization fails.
3. I can write those functions, but I would appreciate some guidance about code architecture as and when needed.

**Comment by timbray** [comment]
Hmm, the key thing you would need to serialize is the coreMatcher type. It references other types: fieldMatcher, valueMatcher, faState, faNext, and smallTable.  There are no generics, they're all struct types. 

What might be a problem is that coreMatcher, fieldMatcher, and valueMatcher use an atomic.Value to support concurrency. I have no idea how that might interact with serialization.

**Reply by timbray** [reply]
Hmm, for example, look at coreMatcher:
```go
type coreMatcher struct {
	updateable atomic.Value // always holds a *coreFields
	lock       sync.Mutex
}
```

Since the atomic.Value isn't typed, not sure how serialization would work.  If that's a problem, I guess there could be a two-step process where you generate 
 
```go
type serializableCoreMatcher struct {
  fields *coreFields
}
```

Easy to serialize, so you have a two-step: Turn coreMatcher into serializableCoreMatcher, then serialize. On the return trip, deserialize into serializableCoreMatcher and use that to create a new coreMatcher. It might mean you have to traverse the whole structure…

**Reply by timbray** [reply]
This is starting to sound complicated. It. might be easier, for the situation where you have a static pattern-matcher with a few hundred rules, to make a ruleBundle type or some such, that serializes all the rules into a compacted form. Then you could retrieve that and rebuild the instance, probably quite efficiently. We could envision storing them in CBOR rather than JSON, or simply just compressing them with the standard `compress` package.

**Reply by markandpathak** [reply]
Same here, as we go deeper in the code, it comes to the point that we need to write serializer for majority of structs. To save time, We came up with exactly the same solution to club all rules and compress them into a single binary blob. We will keep working on the serialization part. Will let you know as we progress.

**Comment by timbray** [comment]
And to make things worse, I'm currently working intensely on refactoring the Quamina data structures :) 

So I think you're taking the right approach.    Ouch, I forgot to mention something that is highly relevant. Check out https://github.com/timbray/quamina/blob/main/README-pruner.md - @jsmorph has already built a system for remembering which patterns have been added to Quamina. This is for the purpose of supporting deletion, but it seems plausible that those remembered patterns could be exported and used to prepare a new Quamina instance.
